services/src/main/java/org/keycloak/services/resources/account/SessionResource.java

```diff
@@ -0,0 +1,222 @@
+
+        for (UserSessionModel s : sessions) {
+            SessionRepresentation rep = new SessionRepresentation();
+            rep.setId(s.getId());
+            rep.setIpAddress(s.getIpAddress());
+            rep.setStarted(s.getStarted());
+            rep.setLastAccess(s.getLastSessionRefresh());
+            rep.setExpires(s.getStarted() + realm.getSsoSessionMaxLifespan());
+            rep.setClients(new LinkedList());
+
+            if (currentSession.getId().equals(s.getId())) {
+                rep.setCurrent(true);
+            }
+
+            for (String clientUUID : s.getAuthenticatedClientSessions().keySet()) {
+                ClientModel client = realm.getClientById(clientUUID);
+                ClientRepresentation clientRep = new ClientRepresentation();
+                clientRep.setClientId(client.getClientId());
+                clientRep.setClientName(client.getName());
+                rep.getClients().add(clientRep);
+            }
+
+            reps.add(rep);
+        }
+
+        return Cors.add(request, Response.ok(reps)).auth().allowedOrigins(auth.getToken()).build();
+    }
+
+    /**
+     * Get device activity information based on the active sessions.
+     *
+     * @return
+     */
+    @Path("devices")
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @NoCache
+    public Response devices() {
+        Map<String, DeviceRepresentation> reps = new HashMap<>();
+        List<UserSessionModel> sessions = session.sessions().getUserSessions(realm, user);
+        UserSessionModel currentSession = auth.getSession();
+
+        for (UserSessionModel s : sessions) {
+            DeviceRepresentation device = DeviceActivityManager.getCurrentDevice(s);
+
+            if (device != null) {
+                DeviceRepresentation deviceRep = reps
+                        .computeIfAbsent(device.getOs() + device.getOsVersion(), s1 -> {
+                            DeviceRepresentation representation = new DeviceRepresentation();
+                            
+                            representation.setLastAccess(device.getLastAccess());
+                            representation.setOs(device.getOs());
+                            representation.setOsVersion(device.getOsVersion());
+                            representation.setDevice(device.getDevice());
+                            representation.setMobile(device.isMobile());
+                            
+                            return representation;
+                        });
+
+                if (deviceRep.getCurrent() == null || Boolean.FALSE.equals(deviceRep.getCurrent())) {
+                    deviceRep.setCurrent(s.getId().equals(currentSession.getId()));
+                }
+                
+                if (Boolean.TRUE.equals(deviceRep.getCurrent())) {
+                    deviceRep.setBrowser(device.getBrowser());
+                    deviceRep.setIpAddress(device.getIpAddress());
+                }
```

Comment:
What does it mean for the device to have a "browser".  I would expect the browser to belong only to the session and not the device.

Also, I'm not sure we have clarified what it means for the device to have a single IP address.  Right now in the UI, I consider the device IP to be the IP from the first session in the list.  That is because this code only sets the IP for the current device, so I can't count on it being present for all devices.

Reply:
The idea is that both IP and browser are available for the current device only. So you can easily check what is the browser and IP currently in use.

If you want to remove both browser and ip from device response, I have no objection.